actions/src/sema4ai/actions/agent/__init__.py

```diff
@@ -0,0 +1,204 @@
+        raise RuntimeError("No invocation context found in the action context!")
+    if not isinstance(invocation_context, dict):
+        raise RuntimeError("invocation_context is not a dictionary!")
+
+    id_value = invocation_context.get(id_key)
+    if not id_value:
+        raise RuntimeError(f"No {id_key} found in the invocation_context!")
+    if not isinstance(id_value, str):
+        raise RuntimeError(
+            f"{id_key} is not a string (found: {id_value} ({type(id_value)}))!"
+        )
+    return id_value
+
+
+def _get_id_from_invocation_context(
+    invocation_context_object, id_key: str
+) -> str | None:
+    invocation_context = invocation_context_object.value
+    if not invocation_context:
+        return None
+    if not isinstance(invocation_context, dict):
+        raise RuntimeError("invocation_context is not a dictionary!")
+
+    # If we have an invocation context, we should always have the id!
+    id_value = invocation_context.get(id_key)
+
+    if not id_value:
+        raise RuntimeError(f"No {id_key} found in the invocation_context!")
+
+    if not isinstance(id_value, str):
+        raise RuntimeError(
+            f"{id_key} is not a string (found: {id_value} ({type(id_value)}))!"
+        )
+    return id_value
+
+
+def _get_id_generic(id_key: str, header_key: str | None = None) -> str:
+    from sema4ai.actions._action import get_current_requests_contexts
+
+    id_value = None
+    request_contexts = get_current_requests_contexts()
+    if request_contexts is None:
+        raise RuntimeError(f"Unable to get the {id_key} (no context available)")
+
+    # Prefer invocation context.
+    if request_contexts.invocation_context:
+        id_value = _get_id_from_invocation_context(
+            request_contexts.invocation_context, id_key
+        )
+
+    if id_value is None and header_key:
+        # Alternative use case when coupled directly with the
+        # agent server without a router to make the invocation/action context
+        # available.
+        request = request_contexts._request
+        if request:
+            id_value = request.headers.get(header_key)
+
+    if id_value is None:
+        action_context = request_contexts.action_context
+        if action_context:
+            # Old use case (without separate id in the action context)
+            id_value = _get_id_from_action_context(action_context, id_key)
+
+    if id_value is None:
+        # Ok, unable to get the id through any of the heuristics,
+        # let's raise an error.
+        raise RuntimeError(f"Unable to get the {id_key}")
+
+    return id_value
+
+
+def get_thread_id() -> str:
+    """Get the thread ID from the action context or the request headers."""
+    return _get_id_generic("thread_id", "x-invoked_for_thread_id")
+
+
+def get_agent_id() -> str:
+    """Get the agent ID from the action context or the request headers."""
+    return _get_id_generic("agent_id", "x-invoked_for_agent_id")
```

Comment:
This needs better docs to say that it'll raise an error if the thread id cannot be found (which is expected when running in VSCode unless a `x-invoked_for_agent_id` is passed as a header in the request in the configured input).